src/ApiConnectors/BaseApiConnector.php

```diff
@@ -82,14 +92,68 @@ public function sendXmlDocument(\DOMDocument $document) {
                 if ($this->numRetries > self::MAX_RETRIES) {
                     break;
                 }
+
+                $this->logRetry($exception);
                 return $this->sendXmlDocument($document);
             }
 
             $this->numRetries = 0;
+            $this->logFailedRequest($exception);
             throw new Exception($exception->getMessage(), 0, $exception);
         }
     }
 
+    private function logSendingDocument(\DOMDocument $document): void
+    {
+        if (!$this->logger) {
+            return;
+        }
+
+        $message = sprintf(
```

Comment:
I think the following would be slightly more clear:
```
$message = 'Sending request to Twinfield.';
if ($this->numRetries > 0) {
    $message .= ' (attempt ' . ($this->numRetries + 1) . ')';
}
```

Take whatever feels best to you.

Reply:
Changed